oauth-common/src/main/java/io/strimzi/kafka/oauth/validator/BackOffTaskScheduler.java

```diff
@@ -0,0 +1,151 @@
+     *
+     * That is to prevent piling up of tasks.
+     *
+     * @return true if task was scheduled for execution, false otherwise
+     */
+    public boolean scheduleTask() {
+        // Only one scheduled task can be outstanding at any time
+        if (!taskScheduled.getAndSet(true)) {
+            log.debug("Acquired taskSchedule lock");
+
+            // First repetition is immediate but at least minPauseSeconds has to pass since the last attempt
+            long delay = 1;
+            long now = System.currentTimeMillis();
+            long boundaryTime = minPauseSeconds > 0 ? lastExecutionAttempt + minPauseSeconds * 1000L : now;
+            if (boundaryTime > now) {
+                delay = boundaryTime - now;
+            }
+            service.schedule(new RunnableTask(), delay, TimeUnit.MILLISECONDS);
+            if (log.isDebugEnabled()) {
+                log.debug("Task scheduled for execution in " + delay + " milliseconds");
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Update the time of last execution attempt to current time.
+     * This is handy when multiple schedulers work the same task on a single thread.
+     * It affects
+     */
+    public void updateLastExecutionTime() {
+        lastExecutionAttempt = System.currentTimeMillis();
+    }
+
+
+    class RunnableTask implements Runnable {
+
+        private int repeatCount = 0;
+
+        @Override
+        public void run() {
+            try {
+                lastExecutionAttempt = System.currentTimeMillis();
+                repeatCount += 1;
+
+                // Delegate to task's run()
+                task.run();
+
+                // Release taskSchedule lock
+                taskScheduled.set(false);
+                log.debug("Released taskSchedule lock");
+
+            } catch (Throwable t) {
+                log.error("Scheduled task execution failed:", t);
+
+                // If things went wrong, reschedule next repetition
+                // in exponential backoff fashion (1,2,4,8,16,32)
+                int delay = (int) Math.pow(2, repeatCount);
+                if (minPauseSeconds > 0 && delay < minPauseSeconds) {
+                    delay = minPauseSeconds;
+                }
+
+                // Limit the delay to maxIntervalSeconds
+                // That makes it possible to grow delay exponentially to some maximum, and then keep it constant
+                if (maxIntervalSeconds > 0 && delay > maxIntervalSeconds) {
+                    delay = maxIntervalSeconds;
+                }
+
+                // Only reschedule if the next run would happen within cutoffIntervalSeconds
+                // If there is another periodic job then we may want to stop re-scheduling this one
+                // once the other job's period is reached.
+                if (cutoffIntervalSeconds <= 0 || delay < cutoffIntervalSeconds) {
```

Comment:
Yes, in this case the execution ended with exception and we are rescheduling thus, we're still 'holding' the task execution. It will be released when run() method completes without exception or when we give up rescheduling.